sql-plugin/src/main/scala/org/apache/spark/rapids/hybrid/CoalesceConvertIterator.scala

```diff
@@ -0,0 +1,165 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.rapids.hybrid
+
+import ai.rapids.cudf.NvtxColor
+import com.nvidia.spark.rapids.{GpuColumnVector, GpuMetric, GpuSemaphore, NvtxWithMetrics}
+import com.nvidia.spark.rapids.Arm._
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import com.nvidia.spark.rapids.hybrid.{CoalesceBatchConverter => NativeConverter}
+import com.nvidia.spark.rapids.hybrid.RapidsHostColumn
+
+import org.apache.spark.TaskContext
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.vectorized.{ColumnarBatch, ColumnVector}
+
+/**
+ * The Iterator wrapper of the underlying NativeConverter which produces the coalesced Batch of
+ * HostColumnVectors. The iterator produces RapidsHostColumn instead of HostColumnVector for
+ * carrying metadata about [Pinned|Pageable]MemCpy which are displayed as Spark SQL Metrics.
+ */
+class CoalesceConvertIterator(cpuScanIter: Iterator[ColumnarBatch],
+                              targetBatchSizeInBytes: Long,
+                              schema: StructType,
+                              metrics: Map[String, GpuMetric])
+  extends Iterator[Array[RapidsHostColumn]] with Logging {
+
+  private var converterImpl: NativeConverter = _
+
+  private var srcExhausted = false
+
+  private val converterMetrics = Map(
+    "C2COutputSize" -> GpuMetric.unwrap(metrics("C2COutputSize")))
+
+  override def hasNext(): Boolean = {
+    // isDeckFilled means if there is unconverted source data remained on the deck.
+    // hasProceedingBuilders means if there exists working target vectors not being flushed yet.
+    val selfHoldData = Option(converterImpl).exists { c =>
+      c.isDeckFilled || c.hasProceedingBuilders
+    }
+    // Check the srcExhausted at first, so as to minimize the potential cost of unnecessary call of
+    // prev.hasNext
+    lazy val upstreamHoldData = !srcExhausted && cpuScanIter.hasNext
+    // Either converter holds data or upstreaming iterator holds data.
+    if (selfHoldData || upstreamHoldData) {
+      return true
+    }
+
+    if (!srcExhausted) {
+      srcExhausted = true
+    }
+    // Close the native Converter and dump column-level metrics for performance inspection.
+    Option(converterImpl).foreach { c =>
+      // VeloxBatchConverter collects the eclipsedTime of C2C_Conversion by itself.
+      // Here we fetch the final value before closing it.
+      metrics("C2CTime") += c.eclipsedNanoSecond
+      // release the native instance when upstreaming iterator has been exhausted
+      val detailedMetrics = c.close()
+      val tID = TaskContext.get().taskAttemptId()
+      logInfo(s"task[$tID] CoalesceNativeConverter finished:\n$detailedMetrics")
```

Comment:
nit: does this need to be at the info level?

Reply:
https://github.com/NVIDIA/spark-rapids/issues/11966